### ISSUE TITLE
Enumerate with model parameters

### ISSUE THREAD

**Issue by null-a** [probmods/webppl#789]
Consider the following, which does MLE for the parameter `p`:

```js
Optimize({
  steps: 10000,
  model() {
    var dist = Bernoulli({p: Math.sigmoid(param({name: 'p'}))});
    observe(dist, true);
    observe(dist, false);
    observe(dist, false);
  }
});
Math.sigmoid(param({name: 'p'})); // => 0.33
```

Intuitively, it seems reasonable to expect that we can replace the `Bernoulli` distribution with a marginal distribution computed by `Enumerate`, without changing the result. Like so:

```js
Optimize({
  steps: 10000,
  model() {
    var dist = Enumerate(function() {
        sample(Bernoulli({p: Math.sigmoid(param({name: 'p'}))}));
    });
    observe(dist, true);
    observe(dist, false);
    observe(dist, false);
  }
});
Math.sigmoid(param({name: 'p'})); // => ??
```

However, this _does not_ compute the same thing as the first example at present. The problem is that `param` will only "notify" `Optimize` about parameters encountered during the evaluation of a model when `Optimize` is the current coroutine. Here, `Enumerate` is the current coroutine, so `param` assumes that we're not doing optimization, and `Optimize` is not aware of these parameters. I think there's a simple fix that will make this particular example work -- instead of only checking the current coroutine we can walk the entire coroutine stack looking for `Optimize`. This also seems like a reasonable thing to do in general.

(Katherine actually ran into essentially this problem, where `Enumerate` came from a RSA sub-model, and the parameters belong to a neural net.)

I think there are a couple of further problems related to this:

1. If `name` is omitted from `param`, we default to using something like _"the bit of the stack address that was added since entering the current coroutine"_ as the parameter name. That's a problem here, since names will be relative to `Enumerate` and not `Optimize`, inducing unexpected parameter sharing. The most direct fix I've thought of for this particular example is to base these relative stack addresses on the inner most coroutine, ignoring `Enumerate`. (i.e. Walk the coroutine stack until we find something that isn't `Enumerate`.) But this seems kinda odd, and might not be the right thing to do in general.
2. We can't swap `param` for `modelParam` in the example above. The problem is that `modelParam` will try to sample from the (continuous) prior from within `Enumerate`, which will generate an error since `Enumerate` only works with discrete choices. I'm not sure what to do about this yet.

**Comment by null-a** [comment]
> We can't swap param for modelParam in the example above.
> I'm not sure what to do about this yet.

Perhaps the fact that using `modelParam` inside of `Enumerate` generates an error here is a good thing. It tells us that we need to move `modelParam` outside of `Enumerate` so that we optimize over it, rather than sum it out. Similarly, perhaps the awkwardness of using `param` directly in the model from within `Enumerate`, is another argument against using `param` directly in the model.

One consequence of adopting this view would be that we would need to re-jig webppl-nn a little. At present, the call to `param`/`modelParam` happens when you apply the function representing the net, which makes it tricky to use the net inside `Enumerate` while calling `param`/`modelParam` outside.

**Comment by ngoodman** [comment]
i think this issue is super important to resolve, since nesting enumerate inside variational is an important use case (and selling point)!

> Perhaps the fact that using modelParam inside of Enumerate generates an error here is a good thing. It tells us that we need to move modelParam outside of Enumerate so that we optimize over it, rather than sum it out. Similarly, perhaps the awkwardness of using param directly in the model from within Enumerate, is another argument against using param directly in the model.

yes, i agree with this view... in the fully bayesian setup the model vars should be sampled outside of the enumerate. of course there is a practical question of whether we need to support some version of "direct param" for ease of use?

> One consequence of adopting this view would be that we would need to re-jig webppl-nn a little. At present, the call to param/modelParam happens when you apply the function representing the net, which makes it tricky to use the net inside Enumerate while calling param/modelParam outside.

right! this interacts with your issues with webppl-nn too, perhaps?

**Comment by null-a** [comment]
> of course there is a practical question of whether we need to support some version of "direct param" for ease of use?

This is the key issue I think. If we do support `param` here (with the implementation I suggested), then webppl-nn works as is, if not then webppl-nn needs tweaking. My initial reaction is that we shouldn't make `param` work here (because it makes `param` harder to understand without buying us much in the way of convenience) but I'm not very confident about this. As a first step, I'll think through the consequences of each approach and share some notes so we have a clearer idea what we're dealing with.

**Comment by null-a** [comment]
Looking at this again, it seems to me that we need to state more clearly what the use of `param` outside of a guide means, in general. (Maybe this is clear to everyone else, but it isn't to me.) I see two broad approaches:

The easiest answer is to say that it doesn't mean anything, i.e. that `param` can only be used in the guide.  However, as we've already touched upon, this would be less convenient to use than what we have now (when it works), so if we go this way, we need to think about how we might retain this convenience. I started to think about this -- there are a few notes [here](https://gist.github.com/null-a/d1d7914d24423a02d66e9f642061742a#towards-a-slightly-more-principled-implementation).

Alternatively, we might figure out what it means to make "model parameters" an actual thing. It's possible that my suggestion of "walking the coroutine stack" (above) would fall out of doing this, but without puzzling out the general case, it seems like a hack to me. I'm not sure what this would look like, or whether we need it?
